subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/CacheableTransformTypeAnnotationHandler.java

```diff
@@ -34,10 +35,16 @@
     @Override
     public void validateTypeMetadata(Class<?> classWithAnnotationAttached, TypeValidationContext visitor) {
         if (!TransformAction.class.isAssignableFrom(classWithAnnotationAttached)) {
-            visitor.visitTypeProblem(ERROR,
-                classWithAnnotationAttached,
-                String.format("Cannot use @%s on type. This annotation can only be used with %s types",
-                    getAnnotationType().getSimpleName(), TransformAction.class.getSimpleName()));
+            visitor.visitTypeProblem(problem ->
+                problem.forType(classWithAnnotationAttached)
+                    .reportAs(ERROR)
+                    .withId(ValidationProblemId.INVALID_USE_OF_CACHEABLE_TRANSFORM_ANNOTATION)
+                    .withDescription(() -> String.format("Using %s here is incorrect", getAnnotationType().getSimpleName()))
+                    .happensBecause(() -> String.format("This annotation only makes sense on %s types", TransformAction.class.getSimpleName()))
+                    .documentedAt("validation_problems", "invalid_use_of_cacheable_transform_annotation")
+                    .addPossibleSolution("Remove the annotation")
+            );
```

Comment:
I like the new API/Builder. We probably don't need the `Id` in the first iteration.

Reply:
There are currently only 2 values in the ID: the "undefined" one, and this one. I think we should create one per problem. It makes them easier to search and we could tag tests with the id, so that it's easier to search. The hard work is already done, creating new IDs shouldn't cost much.

At the same time I wonder if we couldn't have conventions, for example to get the documentation links right from the ID. But that we can improve later.